### PR TITLE
Add delete button for domain registrations and mappings

### PR DESCRIPTION
**Pull request by eltongo**
#### Changes proposed in this Pull Request

This PR adds a delete button for on the registered domain and mapped domain screens, so that the user doesn't have to go to the purchases list.

Here's how it appears in the registered domain screen:

Screenshot 2020-03-18 at 6 18 23 PM

Here's how it appears in the mapped domain screen:

domainnnnnnnn

#### Testing instructions

* Ensure the delete button appears like in the screenshots above, for both mapped domains and registered domains
* For non-refundable registered or mapped domains, make sure the delete button will show the respective dialog
* For refundable domains, make sure the button sends you to the appropriate cancel purchase view

### CODE REVIEW

File: client/my-sites/domains/domain-management/edit/domain-types/mapped-domain-type.jsx [hunk 1]
```diff
@@ -48,10 +51,50 @@ class MappedDomainType extends React.Component {
...
+
+		if ( isLoadingPurchase ) {
+			return <VerticalNavItem isPlaceholder />;
+		}
+
+		if ( ! selectedSite || ! purchase ) {
+			return null;
+		}
+
+		const title = translate( 'Delete your domain mapping' );
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/edit/domain-types/mapped-domain-type.jsx]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **34** times:
`translate( 'Delete Domain Mapping' )` *ES Score: 7*

**Comment by fditrapani** [comment]
Thanks for putting this together! Here are some notes from my test:

- The keep my domain button works in both modal windows when deleting a domain without a refund.
- The modal window is different sizes between the two screens. I have an issue open in our backlog to fix this for the old delete flow, if we fix it there later will it get fixed here too?
- I tried this twice, when I delete a domain that I bought on credits, it takes a long time for anything to happen and then I get an error message that reads "504 status code for "POST /me/purchases/14403048/delete". During this time, I get an email saying the domain is deleted and then when I go back to the domain list the domain is still there until I refresh. I have seen this behaviour in production too. 
image
- When I delete a domain, purchased with a credit card, it takes me to the refund screen in the `/me/purchases` section, shows a success message, and is not there when I get back to the domain listing page. It would be nice if this worked quicker but this is how it currently works so I wouldn't prioritize that for right now.
- The chevron on the delete domain button is darker than the rest of them:
image
- I noticed we're using a placeholder while we load the delete button but the copy is the same in both cases — it doesn't look right with the placeholder graphic animating in two spots. Can we load it in a busy state instead? and maybe add the chevron for both so we can avoid it appearing out of nowhere or just disappearing depending on which one we default to?
- When I try delete a domain mapping, it tells me I need to contact support to cancel my business plan.

File: client/my-sites/domains/domain-management/edit/domain-types/style.scss [hunk 2]
```diff
@@ -0,0 +1,7 @@
+.card {
+	.gridicons-trash {
+		margin-top: -2px;
+		margin-right: 4px;
+		vertical-align: middle;
+	}
```
Comment (stephanethomas) [inline client/my-sites/domains/domain-management/edit/domain-types/style.scss]: Do we really need to duplicate [all those rules](https://github.com/Automattic/wp-calypso/blob/d5a5d38bac742109911789babad6af99611f572e/client/me/purchases/remove-purchase/style.scss#L12)? It seems they are already available from `domain.css`:

screenshot

File: client/my-sites/domains/domain-management/edit/domain-types/mapped-domain-type.jsx [hunk 3]
```diff
@@ -36,7 +37,9 @@ import {
 	isFetchingSitePurchases,
 	hasLoadedSitePurchasesFromServer,
 } from 'state/purchases/selectors';
-import { isRechargeable, isExpired } from 'lib/purchases';
+import { isRechargeable, isExpired, isCancelable } from 'lib/purchases';
+import { cancelPurchase } from 'me/purchases/paths';
+import RemovePurchase from 'me/purchases/remove-purchase';
 import ExpiringCreditCard from '../card/notices/expiring-credit-card';
 import ExpiringSoon from '../card/notices/expiring-soon';
```
[the remark this answers is not in the document]
Comment (eltongo) [reply client/my-sites/domains/domain-management/edit/domain-types/mapped-domain-type.jsx]: Yes, good one!

**Comment by fditrapani** [comment]
Thanks for pointing out the font size @stephanethomas. I think we can also drop the icon for now because it draws unnecessary attention to the button as well.

File: client/my-sites/domains/domain-management/edit/domain-types/style.scss (hunk 2 above)
Comment (eltongo) [reply client/my-sites/domains/domain-management/edit/domain-types/style.scss]: It's not ideal but the duplicated style is needed for this:

```
<CompactCard href={ link }>
    <Gridicon icon="trash" />
    { title }
</CompactCard>
```

**Comment by eltongo** [comment]
>     * The modal window is different sizes between the two screens. I have an issue open in our backlog to fix this for the old delete flow, if we fix it there later will it get fixed here too?

Yes, the domain deletion modal is being reused here, so fixing it in one place should fix it here as well.

>     * I tried this twice, when I delete a domain that I bought on credits, it takes a long time for anything to happen and then I get an error message that reads "504 status code for "POST /me/purchases/14403048/delete". During this time, I get an email saying the domain is deleted and then when I go back to the domain list the domain is still there until I refresh. I have seen this behaviour in production too.

Hmm... that means a timeout has occurred. It must be a backend issue but we'll have to investigate it further.

>     * I noticed we're using a placeholder while we load the delete button but the copy is the same in both cases — it doesn't look right with the placeholder graphic animating in two spots. Can we load it in a disabled state instead? and maybe add the chevron for both so we can avoid it appearing out of nowhere or just disappearing depending on which one we default to?

Sounds good.

>     * When I try delete a domain mapping, it tells me I need to contact support to cancel my business plan.

That's weird. I'll do some more testing and see what's going on. Was that purchased with credits or with a credit card?

**Comment by fditrapani** [comment]
Thanks @elton! 

> That's weird. I'll do some more testing and see what's going on. Was that purchased with credits or with a credit card?

Most likely with credits.

**Comment by eltongo** [comment]
> The chevron for the delete button is darker on hover and seems separate from the rest of the park.
> 
> 
> 
> Is that intentional? It makes the delete button stand out.
> 
> 

Not intentional per se. I’m aware of the issue, but since this is going to be temporary I don’t want to waste too much time on minor issues.

**Comment by eltongo** [comment]
> Just to be sure - do we have a way to track if the domain was deleted from this view (as opposed to `/me/purchases`)? That'd be super helpful to track so that we know this is actually working _and_ how popular this is versus `/me/purchases`. I'd be really curious what would be the split.

I haven't added any extra tracking to this, but I'll do that in a separate PR if that's okay with you.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/3082974

Thank you @eltongo for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/3082974).

**Pull request merged**